### DISCUSSION TITLE
Callback error using Sqlalchemy with Dash plotly

### DISCUSSION THREAD

**Discussion by lsdanttas** [sqlalchemy/sqlalchemy#8787] [Usage Questions]
Hey,

I'm working on updating a dashboard code originally made by a co-worker, I'm new at sqlalchemy and sqlite. What I did was to just update it's database, all the rest of the code is untouched.

When I try to run it on debug mode it gives the error below, but the user table exists (apparently). 

```
Callback error updating url_login.pathname

sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) no such table: user
[SQL: SELECT user.id AS user_id, user.username AS user_username, user.email AS user_email, user.password AS user_password 
FROM user 
WHERE user.username IS NULL
 LIMIT ? OFFSET ?]
[parameters: (1, 0)]
(Background on this error at: https://sqlalche.me/e/14/e3q8)
```

The table was created with metadata (code below). The `create_user_table()` is not called anywhere else on the code, so I guess the error makes sense, but what I'm finding weird is that this same dash code (with other data) is running on a server with no errors.  I don't know if the problem is running it locally. 

Where this `User.metadata.create_all(engine)` should be called? Or is the `User_tbl = Table("user", User.metadata)` already creating the table?

```
db = SQLAlchemy()

# Cria tabela com informações do usuário
class User(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    username = db.Column(db.String(15), unique=True)
    email = db.Column(db.String(50), unique=True)
    password = db.Column(db.String(80))

User_tbl = Table("user", User.metadata)

def create_user_table():
    User.metadata.create_all(engine)
```

**Comment by CaselIT** [comment]
Hi,

Not sure I understand the question.
Create_all should be called when you want to create the tables in the db. That will in general depend on the application.

In your case it seems that you will need to create the db locally once, either by calling create_all or by making a copy of an existing db

**Reply by lsdanttas** [reply]
Thanks, I'll try that. But doesn't the `Table()` object already create a table?

**Reply by CaselIT** [reply]
`Table()` will create a table object, but will not issue the corresponding `CREATE TABLE` on the database

**Reply by lsdanttas** [reply]
Hey @CaselIT, I took some time to better understand all the packages I was working with and now the script seems to be correct and pretty much alike with some examples I found online (like [this one](https://levelup.gitconnected.com/how-to-setup-user-authentication-for-dash-apps-using-python-and-flask-6c2e430cdb51)). But still the same error from the original post appears when I run the dash app.

The function to create the table locally give these informations and the string at the end to confirms the table creation:

```
2022-11-16 15:41:36,632 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2022-11-16 15:41:36,632 INFO sqlalchemy.engine.Engine PRAGMA main.table_info("user")
2022-11-16 15:41:36,632 INFO sqlalchemy.engine.Engine [raw sql] ()
2022-11-16 15:41:36,641 INFO sqlalchemy.engine.Engine COMMIT
Tabela criada
```

I can't find many informations about the error.

**Reply by CaselIT** [reply]
Please ensure that the same file is used. IF the table exists there then no error should be raised.
You may also open the sqlite database with a database explorer application to look if the table are as expected

**Reply by lsdanttas** [reply]
I opened through pandas and an explorer, both show the table. What do you mean by ensure the same file is used? The database file?

**Reply by CaselIT** [reply]
> I opened through pandas and an explorer, both show the table. What do you mean by ensure the same file is used? The database file?

Yes. The error above is for a missing table, so it should not happen if the same file that has the table is used also by the application

**Reply by lsdanttas** [reply]
Yeah I checked that too :( 
I'll give it a double look at all of the references, maybe I missed something idk. Thank you for all the help!
